go-controller/pkg/ovn/master.go

```diff
@@ -842,31 +848,122 @@ func (oc *Controller) addNodeAnnotations(node *kapi.Node, hostSubnets []*net.IPN
+	hostSubnets = hostSubnets[:n]
+	// allocate new subnets if needed
+	if config.IPv4Mode && !foundIPv4 {
+		allocatedHostSubnet, err := oc.masterSubnetAllocator.AllocateIPv4Network()
+		if err != nil {
+			return nil, nil, fmt.Errorf("error allocating network for node %s: %v", node.Name, err)
+		}
+		// the allocator returns nil if it can't provide a subnet
+		// we should filter them out or they will be appended to the slice
+		if allocatedHostSubnet != nil {
+			klog.V(5).Infof("Allocating subnet %v on node %s", allocatedHostSubnet, node.Name)
+			allocatedSubnets = append(allocatedSubnets, allocatedHostSubnet)
+			// Release the allocation on error
+			defer func() {
+				if err != nil {
+					klog.Warningf("Releasing subnet %v on node %s: %v", allocatedHostSubnet, node.Name, err)
+					_ = oc.masterSubnetAllocator.ReleaseNetwork(allocatedHostSubnet)
+				}
+			}()
+		}
+	}
+	if config.IPv6Mode && !foundIPv6 {
+		allocatedHostSubnet, err := oc.masterSubnetAllocator.AllocateIPv6Network()
+		if err != nil {
+			return nil, nil, fmt.Errorf("error allocating network for node %s: %v", node.Name, err)
+		}
+		// the allocator returns nil if it can't provide a subnet
+		// we should filter them out or they will be appended to the slice
+		if allocatedHostSubnet != nil {
+			klog.V(5).Infof("Allocating subnet %v on node %s", allocatedHostSubnet, node.Name)
+			allocatedSubnets = append(allocatedSubnets, allocatedHostSubnet)
+		}
+	}
+	// check if we were able to allocate the new subnets required
+	wantedSubnets := expectedHostSubnets - currentHostSubnets
+	if wantedSubnets > 0 && len(allocatedSubnets) != wantedSubnets {
+		return nil, nil, fmt.Errorf("error allocating networks for node %s: %d subnets expected only new %d subnets allocated",
+			node.Name, expectedHostSubnets, len(allocatedSubnets))
 	}
+	hostSubnets = append(hostSubnets, allocatedSubnets...)
+	klog.Infof("Allocated Subnets %v on Node %s", hostSubnets, node.Name)
+	return hostSubnets, allocatedSubnets, nil
+}
 
-	// Node doesn't have a subnet assigned; reserve a new one for it
-	hostSubnets, err := oc.masterSubnetAllocator.AllocateNetworks()
+func (oc *Controller) addNode(node *kapi.Node) ([]*net.IPNet, error) {
+	oc.clearInitialNodeNetworkUnavailableCondition(node, nil)
+	hostSubnets, allocatedSubnets, err := oc.allocateNodeSubnets(node)
 	if err != nil {
-		return nil, fmt.Errorf("error allocating network for node %s: %v", node.Name, err)
+		return nil, err
 	}
-	klog.Infof("Allocated node %s HostSubnet %s", node.Name, util.JoinIPNets(hostSubnets, ","))
-
+	// Release the allocation on error
 	defer func() {
-		// Release the allocation on error
 		if err != nil {
-			for _, hostSubnet := range hostSubnets {
-				_ = oc.masterSubnetAllocator.ReleaseNetwork(hostSubnet)
+			for _, allocatedSubnet := range allocatedSubnets {
+				klog.Warningf("Releasing subnet %v on node %s: %v", allocatedSubnet, node.Name, err)
+				_ = oc.masterSubnetAllocator.ReleaseNetwork(allocatedSubnet)
```

Comment:
Again, if this fails, shouldn't we log an error?

Reply:
well, it seems that the only error raised by this methods is if that the subnet doesn't belong to the range, that can't be possible if we pick it from the allocators.

Reply:
@aojea that is the way the current function is today, but in the future that function could change to potentially return other errors. We should at least log the error to future proof ourselves imo.